Description: Polkit: Correctly handle the user dismissing the
 authentication dialog
 .
 Adapted from elementary/pantheon-agent-polkit#26:
 * Throw Polkit.Error.CANCELLED when the user dismisses the dialog
 * Listen for events in the Cancellable object instead of using it
 .
 Testing:
 * `pkexec` -> click cancel on dialog -> pkexec exits as dismissed
 * `pkexec` -> Ctrl-C in terminal -> authentication dialog closes
 .
 Fixes #1279
Author: LulzFTW <LulzFTW@users.noreply.github.com>
Origin: https://github.com/solus-project/budgie-desktop/pull/1746
Last-Update: 2019-02-10
---
 src/polkit/polkitdialog.vala | 16 +++++++++++-----
 vapi/polkit-agent-1.vapi     |  2 +-
 2 files changed, 12 insertions(+), 6 deletions(-)

diff --git a/src/polkit/polkitdialog.vala b/src/polkit/polkitdialog.vala
index ed972071..9877f99c 100644
--- a/src/polkit/polkitdialog.vala
+++ b/src/polkit/polkitdialog.vala
@@ -34,6 +34,8 @@ public class AgentDialog : Gtk.Window
     [GtkChild]
     private Gtk.Label? label_error;
 
+    public bool is_cancelled;
+
     public PolkitAgent.Session? pk_session = null;
     private Polkit.Identity? pk_identity = null;
 
@@ -92,6 +94,8 @@ public class AgentDialog : Gtk.Window
         window_position = Gtk.WindowPosition.CENTER_ALWAYS;
 
         key_release_event.connect(on_key_release);
+
+        cancellable.cancelled.connect(on_agent_cancelled);
     }
 
     bool on_key_release(Gdk.EventKey key)
@@ -127,7 +131,7 @@ public class AgentDialog : Gtk.Window
     {
         /* Not authed */
         set_sensitive(true);
-        if (!authorized || cancellable.is_cancelled()) {
+        if (!authorized) {
             label_error.set_text(_("Authentication failed"));
             /* TODO: Cancel non existent spinner */
             var session = pk_session;
@@ -266,9 +270,7 @@ public class AgentDialog : Gtk.Window
         if (pk_session != null) {
             pk_session.cancel();
         }
-        if (!cancellable.is_cancelled()) {
-            cancellable.cancel();
-        }
+        is_cancelled = true;
         done();
     }
 
@@ -285,7 +287,7 @@ public class Agent : PolkitAgent.Listener
     private Budgie.ThemeManager theme_manager;
 
     public override async bool initiate_authentication(string action_id, string message, string icon_name,
-        Polkit.Details details, string cookie, GLib.List<Polkit.Identity?>? identities, GLib.Cancellable cancellable)
+        Polkit.Details details, string cookie, GLib.List<Polkit.Identity?>? identities, GLib.Cancellable cancellable) throws Polkit.Error
     {
 
         var dialog = new AgentDialog(action_id, message, "dialog-password-symbolic", cookie, cancellable);
@@ -305,6 +307,10 @@ public class Agent : PolkitAgent.Listener
 
         dialog.destroy();
 
+        if (dialog.is_cancelled) {
+            throw new Polkit.Error.CANCELLED("Authentication dialog was dismissed by the user");
+        }
+
         return true;
     }
 
diff --git a/vapi/polkit-agent-1.vapi b/vapi/polkit-agent-1.vapi
index 7183ea77..d8f402a8 100644
--- a/vapi/polkit-agent-1.vapi
+++ b/vapi/polkit-agent-1.vapi
@@ -6,7 +6,7 @@ namespace PolkitAgent {
 	public abstract class Listener : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected Listener ();
-        public virtual async bool initiate_authentication(string action_id, string message, string icon_name, Polkit.Details details, string cookie, GLib.List<Polkit.Identity?>? identities, GLib.Cancellable cancellable);
+        public virtual async bool initiate_authentication(string action_id, string message, string icon_name, Polkit.Details details, string cookie, GLib.List<Polkit.Identity?>? identities, GLib.Cancellable cancellable) throws Polkit.Error;
         /*
         public virtual void initiate_authentication(string action_id, string message, string icon_name, Polkit.Details details, string cookie, GLib.List<Polkit.Identity?>? identities, GLib.Cancellable cancellable, GLib.AsyncReadyCallback @callback);
 		public virtual bool initiate_authentication_finish (GLib.AsyncResult res) throws GLib.Error;*/
-- 
2.19.1

